### PR TITLE
`template.Rmd` gains case-handling for input scatter data with missing entries

### LINKED ISSUE #59: scatter plot not working
it seems like there is an issue with putting the three parts of the scatter plot together, but my `ggplot2` knowledge is too basic to understand the issue quickly. I can look into this in case you have no time @MonikaFu 
Otherwise I think it makes more sense if you look into this

This is the error that is returned in the last bit of `plot_scatter()`, where all three pieces are put together
```
Error in `[[<-.data.frame`(`*tmp*`, i, value = c(16, 1, 16, 16)) : 
  replacement has 4 rows, data has 30
```

### PR DESCRIPTION
**Pull request by jdhoffa**
Here, I have added a simple test in the scatterplot plotting functions, which test if the input data has ALL expected entity_types (namely "portfolio", "peer", "peers_mean", and "benchmark"). 

If it doesn't have these groups (or in fact, if it has any more or less than exactly these groups, the function won't know what to plot). 

Note: @MonikaFu all this does is prevent the scatterplots from being plotted if there isn't useful data. However, the entire text chunk will still appear, which might be confusing... not sure how you want to handle that 

Closes #59

### CODE REVIEW

File: inst/extdata/PA2022CH_en_exec_summary/template.Rmd [hunk 1]
```diff
@@ -185,7 +192,7 @@ data_scatter_power_e <- prep_scatter(
 The aggregated score compares the alignment of all assets until 2026 to the [XYZ] Scenario. The score is calculated both on aggregate portfolio level (see also Climate Scores page 10) and per PACTA sector (except for cement).
 
 ```{r scores, fig.show='hold', out.width='50%', eval = FALSE}
-# FIXME: turn eval back on
+
```
Comment (cjyetman) [inline inst/extdata/PA2022CH_en_exec_summary/template.Rmd:195]: probably a mistake

**Review by MonikaFu — CHANGES_REQUESTED** [review]
I'd rather put the check inside of `check_data_scatter()` in `plot_scatter()`. Seems like a more robust place to put it. I can also add a check for the actual values for `entity_type` that we're expecting.

**Comment by jdhoffa** [comment]
> I'd rather put the check inside of `check_data_scatter()` in `plot_scatter()`. Seems like a more robust place to put it. I can also add a check for the actual values for `entity_type` that we're expecting.

This sorta relates to a parallel conversation going on here: https://github.com/RMI-PACTA/pacta.portfolio.analysis/issues/17

The gist of it is: What should we do when a function is passed data (with the right format) that it can't do anything with (e.g. an empty data-frame)? 

Do we: 
1.  Cause the function to error, saying "I can't do anything with this dataset" 
2.  Pass a warning, and return an empty tibble (or empty plot or something similar) 

1. Makes a lot of sense, because it seems sorta logical that a function would just error if it is passed weird data, however then we need to think of some way to achieve error handling, or do checks BEFORE the data is even passed to the function (this would occur in whatever workflow script is calling the function)
2. Also makes sense, because it allows all of our workflow scripts to pass without any errors, and consider all possible cases, however it might mean functions have some semi-unexpected behaviour, (what exactly would you even want a function to return if it is passed incorrect data). 

Personally my preference would be 1, but honestly I don't think we've come to a conclusion on what approach to use.

**Comment by jdhoffa** [comment]
> Do we:
> 
> 1. Cause the function to error, saying "I can't do anything with this dataset"
> 2. Pass a warning, and return an empty tibble (or empty plot or something similar)
> Option 1 Makes a lot of sense, because it seems sorta logical that a function would just error if it is passed weird data, however then we need to think of some way to achieve error handling, or do checks BEFORE the data is even passed to the function (this would occur in whatever workflow script is calling the function)
> Option 2 Also makes sense, because it allows all of our workflow scripts to pass without any errors, and consider all possible cases, however it might mean functions have some semi-unexpected behaviour, (what exactly would you even want a function to return if it is passed incorrect data).
> 
> Personally my preference would be 1, but honestly I don't think we've come to a conclusion on what approach to use.

Ah, I've just noticed that in #112 you decided to go with Option 2 hahaha. Then let's go with that approach now, but definitely I think this is something we should consider, and align all `pacta.*` packages on in the future, so we make sure to only have to do case-handling in one place (either packages, or workflow scripts)

**Comment by jdhoffa** [comment]
It seems that this edge-case is not in fact handled in #112, so I will re-open this PR

**Comment by MonikaFu** [comment]
Indeed #112 is handling a different case in which there is no data for an asset class in a portfolio (coming from `pacta.portfolio.analysis` results). In that specific case I do want `data_prep_*()` to return an empty dataset which is then passed to the plotting function and triggers returning an empty plot with a message that no data was found. 

The case that you are touching upon here is different. There is data for portfolio but peers or indices are missing. In that case `data_prep_scatter()` will return a dataset but it will not contain all the `entity_type` values for the plot to be created. `plot_scatter` required 4 entity_type values to be present in the data: c("this_portfolio", "peers", "peers_mean", "benchmark") otherwise it will crash. In theory I could make the plot more robust and not cause it to crash if some of those are missing but in practice if all the data are in place, these 4 entity types should be there so I'd say that it is fair for the plot to return an error if it doesn't find them in the input.

**Comment by MonikaFu** [comment]
@jdhoffa I am sorry to mess with your PR but I added a check on the plot side. I think that we could still keep your implementation in the template if we want the template to still render even though there is something wrong with the data in this plot. I don't really have an opinion on this. @AlexAxthelm ?

**Comment by AlexAxthelm** [comment]
I'm not totally sure who's testing for what and where, but my inclination is to say that functions should check their own assumptions.

I think this also aligns with my thoughts on:

> The gist of it is: What should we do when a function is passed data (with the right format) that it can't do anything with (e.g. an empty data-frame)?

Keeping this as a big question, I think there's an important difference between what we do as part of analysis/calculation/prep layer vs as part of the presentation layer. So it would be appropriate (for example) the analysis to pass along empty objects as part of a calculation pipeline (dropping warnings where appropriate), but when we get to presentation, to not render (or render the new `empty_plot` object).

**Comment by jdhoffa** [comment]
> @jdhoffa I am sorry to mess with your PR but I added a check on the plot side. I think that we could still keep your implementation in the template if we want the template to still render even though there is something wrong with the data in this plot. I don't really have an opinion on this. @AlexAxthelm ?

Yeah so this is the point I was trying to make... the solution you just added to this PR now actually conflicts conceptually with the solution in #112. 

In #112, if empty data is passed, the function does not error, it just passes an empty data-frame. 
In #113, if data missing certain entries is passed, the function errors (meaning that there needs to also be a check in .Rmd to ensure the error doesn't block rendering). 

Since empty data is just a special case of data missing certain entries (ie. missing all entries), it is strange that we would have two different behaviours. To be clear, I agree that functions should test their own conditions. But if failure of that condition causes a function to error, then we need to have a second check on the implementation side (e.g. here in the .Rmd). 

In this situation, we need to have the .Rmd check ONE condition (ie that the data has 4 entity_types), but not all of the conditions (ie. that the data isn't empty), because some of the conditions are handled different within the function. 

I am not fussed about how we do this, but we just need to try to be consistent.

**Comment by jdhoffa** [comment]
@AlexAxthelm @MonikaFu is anyone against me softening the behaviour in this PR from an error to a warning, and passing an empty_plot (as in #112)?

**Comment by MonikaFu** [comment]
@jdhoffa as we just discussed in the meeting. My preference would be for the function to error and have an error handling functionality introduced to handle this case (see #112). The case that we're looking at here is an actual error and not missing data issue. Missing data for a portfolio would cause the plot to return a 'No data' plot. If that was not the case, it means that indices or peer data are missing which I'd say is an error.

**Comment by jdhoffa** [comment]
> @jdhoffa as we just discussed in the meeting. My preference would be for the function to error and have an error handling functionality introduced to handle this case (see #112). The case that we're looking at here is an actual error and not missing data issue. Missing data for a portfolio would cause the plot to return a 'No data' plot. If that was not the case, it means that indices or peer data are missing which I'd say is an error.

Ok, then I think this is good to merge as is (though it will result in no plots right now). 
And then once #121 is addressed, we can replace the "no plots", with an "error plot". 

Sound good?

**Comment by jdhoffa** [comment]
Maybe closes #59

**Review by cjyetman — APPROVED** [review]

**Review by MonikaFu — APPROVED** [review]
LGTM

**Pull request merged**